File: java/src/main/java/io/aurigraph/dlt/sdk/namespace/WalletApi.java
package io.aurigraph.dlt.sdk.namespace;

import com.fasterxml.jackson.core.type.TypeReference;
import io.aurigraph.dlt.sdk.AurigraphClient;
import io.aurigraph.dlt.sdk.dto.Transaction;
import io.aurigraph.dlt.sdk.dto.TransferReceipt;
import io.aurigraph.dlt.sdk.dto.TransferRequest;
import io.aurigraph.dlt.sdk.dto.WalletBalance;

import java.net.URLEncoder;
import java.nio.charset.StandardCharsets;
import java.util.List;

/**
 * Namespace for /api/v11/wallet endpoints — wallet management and transfers.
 */
public class WalletApi {

    private final AurigraphClient client;

    public WalletApi(AurigraphClient client) {
        this.client = client;
    }

    /**
     * Get the balance for a wallet address.
     */
    public WalletBalance getBalance(String address) {
        String enc = URLEncoder.encode(address, StandardCharsets.UTF_8);
        return client.get("/wallet/" + enc + "/balance", WalletBalance.class);
    }

    /**
     * Transfer tokens between wallets.
     */
    public TransferReceipt transfer(TransferRequest req) {
        return client.post("/wallet/transfer", req, TransferReceipt.class);
    }

    /**
     * Get transaction history for a wallet address.
     */
    public List<Transaction> getHistory(String address) {
        String enc = URLEncoder.encode(address, StandardCharsets.UTF_8);
        return client.get("/wallet/" + enc + "/history", new TypeReference<List<Transaction>>() {});
    }
}

File: android/aurigraph-sdk/src/main/kotlin/io/aurigraph/sdk/apis/WalletApi.kt
package io.aurigraph.sdk.apis

import io.aurigraph.sdk.AurigraphClient
import io.aurigraph.sdk.models.Transaction
import io.aurigraph.sdk.models.TransferReceipt
import io.aurigraph.sdk.models.TransferRequest
import io.aurigraph.sdk.models.WalletBalance

/**
 * Wallet management and token transfers.
 */
class WalletApi(private val client: AurigraphClient) {

    /** Get the balance for a wallet address. */
    suspend fun getBalance(address: String): WalletBalance =
        client.get("/wallet/$address/balance")

    /** Transfer tokens between wallets. */
    suspend fun transfer(request: TransferRequest): TransferReceipt =
        client.post("/wallet/transfer", request)

    /** Get transaction history for a wallet address. */
    suspend fun getHistory(address: String): List<Transaction> =
        client.get("/wallet/$address/history")
}
